Updated Simulation Tab - BOIN and Plots

christie94h opened the pull request:
This PR is the second draft of a working Simulation tab. 

## Changes 

### BOIN Added to Trial Designs
- The Other in Trial Designs has now been changed to include BOIN inputs. There is a button at the top of these inputs to determine how to input the data (in particular, how to input the escalation/de-escalation probabilities/boundaries). 
- There is a button below the entry points which will be used to generate escalation boundaries if the probabilities are entered.
- A BOIN simulation function was also added, which is used just as the existing 3+3 and CRM ones. 

### Simulation Tab UI

- I have changed the presentation of the Simulation UI to include 3 further tabs: the inputs, table outputs and plot outputs.
- The inputs remain unchanged.
- The table output tab is now split into two cards, which (will in future) display the same tables, one by one as in the previous Simulation tab. The idea here is that the user can scroll down within each card to compare any two of their desired outputs.
- The plot output tab displays plots, either 'by model' or 'by scenario'. What that means is that when comparing by model, the desired models appear on the same plot and the different scenarios correspond to different plots, and similarly for by scenario. There are some pictures below of what the charts should look like.

### Fixing Issues from the First Draft
This PR is the final part of the solution to issue #45. Once this PR is approved I will close that issue.

### The Plot Outputs
- When any simulation is run, plots will output in the Plot Outputs tab, and depending on what the user has selected, these plots will either be 'by model' or 'by scenario'. 
- There are two types of plot: bar charts for the % plots (% dose chosen as MTD and % of patients treated at dose) and histograms for the distributions (accuracy, length, overdose). 
- The bar charts appear like in the image below (which is 'by model').  The crosshatch patten shows the dose that is the true MTD. 
<img width="1433" height="636" alt="image" src="https://github.com/user-attachments/assets/76afab3f-6d15-47e6-951b-36f6627b17ad" />
- The histograms appear like in the image below (which is 'by scenario').  The dotted lines are the mean values (which are quoted in the tables tab). Note that there are issues with the Accuracy plots - this is explained in further detail below.
<img width="1429" height="626" alt="image" src="https://github.com/user-attachments/assets/d1991538-b99a-400a-824f-a209d39fb28e" />

## Unfinished Tasks and Broken Code

- (DONE in this PR) In the Trial Designs tab, there is a button to generate escalation boundaries for the BOIN data. It doesn't do anything currently, so that needs to be added. 
- (DONE in PR #69) The accuracy distribution plots don't actually plot accuracy, but instead a similar plot to % of simulations that chose a dose as the MTD. This needs to be reworked (at the simulation level as the simulation functions themselves are the issue).
- (DONE in PR #69) The tables display rework doesn't work, as the tables do not appear on the right comparison card. How to present the tables needs to be thought about, and if it is decided that we should stick with the current format, then the right comparison card needs to be fixed.
- (NOT DONE) The pattern labels on the bar charts don't work.
- (NOT DONE) A method to download possible plots and tables needs to be added.
-  (NOT DONE) A method to upload and download trial design responses needs to be added.
-  (DONE in PR #71) Validation needs to be added to some of the trial design inputs and all of the simulation inputs.
- (NOT DONE) When writing the 'by Scenarios' part of the plots, I came across a bug that required me to change most of the previous code. As a result, the finished code without the bug is lengthier than it could/should be, so it needs to be cut down.

## Testing
- [ ] Open the app and head to the Trial Designs page. Can you enter variables into the BOIN designs?
- [ ] Move to the Simulations tab and run a simulation with BOIN included, and look at the tables. Then change a BOIN specific input (and general input) and run the simulation again. Do the results change?
- [ ] Run a simulation to output plots by model. Do the plots appear as expected and do they change which data inputs are changed?
- [ ] Run a simulation to output plots by scenario. Do the plots appear as expected and do they change which data inputs are changed?
- [ ] Run simulations for Scenarios which don't include Scenario 1 or include Scenario 3 but not Scenario 2 (e.g. Scenarios 1 and 3). Are the tables and plots titled correctly?

Commit:
Copying the code from boinsim.R into global.R

Commit:
Altering the code within sim_boin to make it a functional function.

Commit:
Coding in the UI for BOIN within Trial Designs.

Commit:
Adding simulation outputs for BOIN using hardcoded simulation inputs.

Commit:
Adding in shared values to BOIN simulation function

Commit:
changing the sim_boin function to match the titles of the simulation inputs

Commit:
Fixing the height of the cards to fit the reactive table

Commit:
Writing a function to plots the results (accuracy is wrong)

Commit:
Writing a data_for_plotting function to help plot data from simulations

Commit:
Setting up a new UI design for the Simulations tab to accommodate plots.

Commit:
Writing a function to plot generic bar charts

Commit:
Updating the plot_bar function to allow for model differentiation

Commit:
Written code in the simulation tab that displays the mtd graph if models and a singular scenario are selected.

Commit:
A graph for % mtd and % treated now appear for Scenario 1 when "By model" is clicked.

Commit:
Added multiple scenarios to the plots (although it is impossible to tell which scenario is which currently)

Commit:
Added histograms to plot output.

Commit:
Simulation plots only display if the metric is selected.

Commit:
Fixed model naming in the plots and removed unnecessary `print` outputs

Commit:
Putting mean lines in the right place on the plots

Commit:
Colouring and placing the mean values on the plots correctly.

Commit:
Adding scenarios to titles of 'by Model' plots

spoofwrm committed:
first conduct tab skeleton (does not work)

Commit:
Set up plots by scenario for fixed mean value.

Commit:
Changing the 'by model' plots to accommodate for change in plot_bar and plot_dist functions

spoofwrm committed:
Have filler figures and tables on conduct tab (updated renv too)

spoofwrm committed:
rendered placement table and graph into conduct tab

Commit:
Fixed several bugs (and temporarily removed mean lines from histograms in order to fix said bugs)

spoofwrm committed:
added reactive table to results card

spoofwrm committed:
reactive DLT values are now a tick box

Commit:
Plots display in the simulations tab when called. There are 4 bugs still to fix with the presentation of these plots.

Commit:
Fixing some bugs that meant labelling in plots was incorrect. Also fixed the incorrect Scenario naming in the tables.

spoofwrm committed:
normal table with dlt corresponding to colours on graph (must type true or false)

Commit:
Adding a pattern to signify the true MTD

Commit:
Deleting unused functions and comments to reduce the size of the the code.

Commit:
Moving code to speed up runtime

Commit:
Adding in a deleted function that I didn't want to delete (oops)

spoofwrm committed:
edited "DLT" to dose_level

Commit:
Generate esclation boundaries in BOIN Trial Design now works.

spoofwrm committed:
added function button, does not have an automatic max_size

Commit:
Added BOIN values to shared variables list

spoofwrm committed:
have add and remove cohort buttons on top of max_size, as well as Dose_level that updates whole cohort when dose level changed

Commit:
Removing dist_accuracy from the functions in global.R (as dist_accuracy is the same graph as % MTD)

Commit:
Removed dist_accuracy from the simulation plots

Commit:
Changed histograms to density plots

Commit:
Coded in Individual and Comparative outputs for the plots

Commit:
Individually displaying tables in the table outputs

spoofwrm committed:
removed attempts to fix show entries (messed it up)

spoofwrm committed:
fixed remove cohort logic

spoofwrm committed:
first draft of the value boxes, but logic works

spoofwrm committed:
design of value boxes are ideal, but logic goes stale for dose level

spoofwrm committed:
fully corrected value boxes and renv update (bsicons)

spoofwrm committed:
given user option to change graph title

spoofwrm committed:
added the first bit of the CRM/BOIN results table

spoofwrm committed:
added recommended dose level value card, that informs the add_cohort button to add new cohort at recommended dose level

spoofwrm committed:
Have reactive card for just crm design and implemented the global shared values

Commit:
Adding individual view to tables

Commit:
Adding an option when there is only one table in comparative view.

spoofwrm committed:
starting to add in logic for rmd generation as well as pdf/excel, not fully fleshed out yet

Commit:
Added in function based on CRM dose-tixocity models to output scenario true DLT vectors.

Commit:
Adding the example scenarios to the reactive table.

spoofwrm committed:
added patient column (with dose level working)

spoofwrm committed:
moved legend to outside graph

Commit:
Added function to initialise a reactive table on app start-up.

spoofwrm committed:
Have a CRM results table with dose level and no. of dlts linked to data table. Posteriors and CI's are placement values till I establish the logic to work them out

spoofwrm committed:
Added BOIN resutls card and table with desirability score (not statistically accurate)

Commit:
Adding basic mode to trial designs tab

Commit:
Changing the aesthetics of the Trial Design tab.

Commit:
Changed the lines in the distribution plots from means to medians

spoofwrm committed:
add boin and crm tables to the download report (pdf still not working but excel is all good)

spoofwrm committed:
bracket move around lol

spoofwrm committed:
rearranging and a renv update

Commit:
Adding validation to BOIN inputs

spoofwrm committed:
shared variables work across app (yippee), need to update add_cohort button to adapt to cohort size

spoofwrm committed:
add cohort button updated to adapt to cohort size

spoofwrm committed:
changed table to hit max_size only, and added value box for trial status

spoofwrm reviewed (commented):
Looks great, I know there are some knots you are working on now but this all looks great so far!

Commit:
Writing a function to stop BOIN generations crashing when incorrect probabilities entered.

Commit:
Correcting the typo in the shared variables list

christie94h commented:
I've fixed a few of the major issues in PR #69 and have requested it be merged into this branch to avoid issues when merging to dev.

Commit:
Adding validation to Basic Mode and removing an unecessary test function

Commit:
Adding validation to numeric simulation inputs

Commit:
Adding validation to the simulation input reactive table

Commit:
Fixing the error of simulations not running without table being refreshed.

xlai committed:
Merge pull request #69 from xlai/fixing-accuracy

Fixing the Issues in #67

Commit:
Change to debug.log

Commit:
Merge branch 'boin-simulations' of https://github.com/xlai/dosefinderhub into boin-simulations

christie94h committed:
Merge branch 'dev' into boin-simulations

Commit:
Rewriting generation of example scenarios to take into account when the MTD is an edge case, or when n_doses is 1.

Commit:
Editing the example scenario generation to allow for cases where the prior MTD is larger than the number of doses (i.e. those using basic mode who haven't changed the prior MTD)

spoofwrm committed:
added in escalation logic for CRM  (still needs some work)

spoofwrm committed:
added in escalation logic for CRM  (still needs some work)

Commit:
Fixing an error in the previous commit

Commit:
Trying (and failing) to fix the questionnaire data bug. (When questionnaire data is moved over, the shared variables get stuck as the questionnaire responses).

Commit:
Using clunky code to fix the questionnaire variables bug (this needs to be changed in future).

christie94h commented:
I believe I have fixed the two major errors outlined in @xlai's comment in PR #69:

1. For the issue with BOIN crashes after simulations, I can't seem to get the app to crash when I try changing inputs. I'm hoping that the issue was fixed when I merged things onto this branch from `dev`, but if not, do let me know and I'll try to fix it again.
2. The issue with the initial table has been fixed.

In trying to fix these issues, I ran into two further issues:

1. The example scenarios weren't defined when there is only one dose and when the MTD is an edge case. This has been resolved and I also added a feature to choose a new MTD if the inputting MTD is larger than the number of doses, to avoid those who went through the basic mode having to go into the advanced mode to fix the general MTD.
2.  When the questionnaire inputs were transferred across to the Trial Designs tab, the reactive nature of the shared variables was lost, and they remained at the questionnaire outputs indefinitely. I have fixed this with a clunky piece of code that effectively forces a change every time a general input is edited. This is not ideal and needs to be fixed (as I've noted in the code), but I have left it for now so that the app is functional to present tomorrow. I have written this problem up as issue #72, so I don't forget to fix it later down the line.

christie94h committed:
Adding validations to simulation tab and block simulation if there is warning (#71)


- Add validation for trial designs and simulation inputs
- Display error messages with tab-specific context
- Ensure table completeness and value constraints (0-1 range, increasing order)
- Require table refresh before running simulations
- Enable navigation to Simulations tab via View Simulations button

Commit:
Updating renv

spoofwrm committed:
Connecting escalation logic to crm results card (doesn't work)

Commit:
Adding in the code to transfer questionnaire variables over to trial design that was lost in a merge.

Commit:
Fixing the crm prior skeleton input.

spoofwrm committed:
Revert "added in escalation logic for CRM  (still needs some work)"

This reverts commit fc9adc5921a0fee686d83b295367b14035185794.

reverting

spoofwrm committed:
Revert "Connecting escalation logic to crm results card (doesn't work)"

This reverts commit a48b74a8a852d4c81c309a8ca7177fe51690e423.

Commit:
A quick (and not permanent) fix to the true DLT examples so the sequence is always increasing.

Commit:
Adding prior mtd to basic mode (without validation)

Commit:
Rewriting the example_scenarios function to take dose into account

Commit:
Updating code in Simulation tab to reflect the new general true DLT formula.

spoofwrm committed:
Updated escalation for all three designs, and update design logic to prompt button pressing

spoofwrm committed:
Posterior tables working for CRM and BOIN (CI's not yet)

Commit:
Fixing the tpt input to describe what it actually measures.

spoofwrm committed:
edits to downloadHandler (doesn't work) and added aesthetics to graph

Commit:
Adding a dropdown list for individual table view and displaying tables individually.

Commit:
Removing allow_deesc from 3+3 designs

spoofwrm committed:
simplified recommendation logic

spoofwrm committed:
added 3+3 for escalation (has issues when a cohort has 2 DLT's)

Commit:
Changing the UI to create a new "basic" and "advanced" mode.

Commit:
Added server logic to basic and advanced modes.

spoofwrm committed:
Added safe stopping for when trial is too toxic, and MTD message in recommended value box

spoofwrm committed:
Add button disabled when trial is stopped

Commit:
Adding logic to ask the user to transfer to advanced mode if view simulations is clicked.

Commit:
Fixing validation in 'view simulations' button

Commit:
Changing comparative viewby model to a dropdown with basic names.

Commit:
Doing the same as the previous commit but for comparisons by scenario.

Commit:
Changing the generic names to match the graph titles

Commit:
Changing all instances of 'model' to 'design'

Commit:
Coding in individual view's bar charts

Commit:
Added individual view of density plots with hard-coded medians

Commit:
Adding correct medians to individual distribution plots

spoofwrm committed:
Marking exp (dev2)

spoofwrm committed:
Merge branch 'exp' into adding-conduct

spoofwrm committed:
Merge pull request #68 from xlai/adding-conduct

Adding Conduct Tab to exp branch from dev

christie94h committed:
Merge branch 'exp' into boin-simulations
